pkg/codegen/go/gen_program_expressions.go

```diff
@@ -740,6 +744,16 @@ func (g *generator) genTupleConsExpression(w io.Writer, expr *model.TupleConsExp
 	}
 	g.genTemps(w, temps)
 	argType := g.argumentTypeName(expr, destType, isInput)
+	// don't need to generate type for list args if not a pointer, i.e. []ec2.SubnetSpecArgs{ {Type: ...} }
+	// unless it contains an interface, i.e. []map[string]interface{ map[string]interface{"key": "val"} }
+	if strings.HasPrefix(argType, "[]") {
+		if !strings.Contains(argType, "interface{}") {
+			g.isGenTupleConExprListArgs = true
+			if strings.HasPrefix(argType, "[]*") {
+				g.isPtrArg = true
+			}
+		}
+	}
```

Comment:
Do we want `*Array` and `*ArrayInput` types (such as [ListenerCertificateArrayInput](https://github.com/pulumi/pulumi-aws/blob/64c83a9e5556bbb17d1f20e12a24614c32a81655/sdk/go/aws/applicationloadbalancing/listenerCertificate.go#L174-L181))? My understanding is that those are still array literals, but they don't start with `[]`.

I don't think setting the values to false at the end will work with nested `genTupleConsExpression` calls. I think we will need to restore these to their previous values after setting them, rather than setting them to `false` later. It would be great if we had a test for nested cons expressions.

```suggestion
	if strings.HasPrefix(argType, "[]") && !strings.Contains(argType, "interface{}") {
		defer func(b bool) { g.isGenTupleConExprListArgs = b } (g.isGenTupleConExprListArgs)
		g.isGenTupleConExprListArgs = true
		if strings.HasPrefix(argType, "[]*") {
			defer func(b bool) { g.isPtrArg = b } (g.isPtrArg)
			g.isPtrArg = true
		}
	}
```

Reply:
I'm not sure if it's valid Go for type aliases – for example, removing `&s3.BucketLoggingArgs` from [this example ](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/testing/test/testdata/aws-s3-logging-pp/go/aws-s3-logging.go#L15-L17) causes an 'invalid composite literal' error

Good point on the nested expression 😄